effekt/shared/src/main/scala/effekt/core/PatternMatchingCompiler.scala

```diff
@@ -179,12 +179,24 @@ object PatternMatchingCompiler {
 
       // used to make up new scrutinees
       val varsFor = mutable.Map.empty[Id, List[ValueVar]]
-      def fieldVarsFor(constructor: Id, fieldTypes: List[ValueType]): List[ValueVar] =
-        varsFor.getOrElseUpdate(constructor, fieldTypes.map { tpe => ValueVar(TmpValue("y"), tpe) })
+      def fieldVarsFor(constructor: Id, fieldInfo: List[((Pattern, ValueType), String)]): List[ValueVar] =
+        varsFor.getOrElseUpdate(
+          constructor,
+          fieldInfo.map {
+            // if it's a pattern named by the user in the program, use the supplied name
+            case ((Pattern.Any(id), tpe),         _) => ValueVar(TmpValue(id.name.name), tpe)
+            // otherwise, use the field name of the given field
+            case ((_,               tpe), fieldName) => ValueVar(TmpValue(fieldName),    tpe)
```

Comment:
```suggestion
            // if it's a pattern named by the user in the program, use the supplied name
            case ((Pattern.Any(id), tpe),         _) => ValueVar(Id(id.name.name), tpe)
            // otherwise, use the field name of the given field
            case ((_,               tpe), fieldName) => ValueVar(Id(fieldName),    tpe)
```